File: src/lib/sammgmtjni/com/sun/netstorage/samqfs/mgmt/adm/License.java
/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2008 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */
// ident	$Id: License.java,v 1.9 2008/05/16 18:35:27 am143972 Exp $

package com.sun.netstorage.samqfs.mgmt.adm;

import com.sun.netstorage.samqfs.mgmt.Ctx;
import com.sun.netstorage.samqfs.mgmt.SamFSException;
import com.sun.netstorage.samqfs.mgmt.media.MdLicense;

/**
 *  this class provides access to SAM-FS/QFS license-related information
 */
public class License {
    // License information
    private short type   = -1;
    private short fsType = -1;
    private long expire  = -1;
    private long hostID  = -1;
    private int featureFlags = -1;   // bit mask of all feature available
    private MdLicense[] mdLicenses;

    /**
     * private constructor (no public constructor needed)
     */
    private License(short type, long expire, short fsType,
        long hostID, int featureFlags, MdLicense [] mdLicenses) {

            this.type = type;
            this.fsType = fsType;
            this.expire = expire;
            this.hostID = hostID;
            this.featureFlags = featureFlags;
            this.mdLicenses = mdLicenses;

    }

    // these constants below must match those in pub/mgmt/license.h
    // license expiration info
    public static final short NON_EXPIRING  = 0; // License never expires
    public static final short EXPIRING = 1; // License expires on exp_date
    public static final short DEMO = 2; // DEMO lic. exp. after 14 days
    public static final short QFS_TRIAL = 3; // License expires after 60days
    public static final short QFS_SPECIAL = 4; // License never expires

    // licensed FS type
    public static final short QFS    = 1; // stand alone QFS
    public static final short SAMFS  = 2; // SAM-FS
    public static final short SAMQFS = 3; // SAM-QFS

    // remote sam server supported
    private static final int REMOTE_SAM_SVR_SUPPORT  = 0x00000001;
    // remote sam client supported
    private static final int REMOTE_SAM_CLNT_SUPPORT = 0x00000002;
    // migration toolkit supported
    private static final int MIG_TOOLKIT_SUPPORT = 0x00000004;
    // fast file system supported
    private static final int FAST_FS_SUPPORT = 0x00000008;
    // database feature supported
    private static final int DATABASE_SUPPORT = 0x00000010;
    // foreign tape supported
    private static final int FOREIGN_TAPE_SUPPORT = 0x00000020;
    // shared file system supported
    private static final int SHARED_FS_SUPPORT = 0x00000040;
    // segment supported
    private static final int SEGMENT_SUPPORT = 0x00000080;
    // SAN API supported
    private static final int SANAPI_SUPPORT = 0x00000100;
    // standalone qfs supported
    private static final int STANDALONE_QFS_SUPPORT = 0x00000200;

    // public methods

    public short getLicenseType() { return type; }

    public long getExpiration() { return expire; }

    public short getFSType() { return fsType; }

    public long getHostID() { return hostID; }

    public boolean isRemoteSamServerSupported() {
        return (((featureFlags & License.REMOTE_SAM_SVR_SUPPORT)
            == License.REMOTE_SAM_SVR_SUPPORT) ? true : false);
    }

    public boolean isRemoteSamClientSupported() {
        return (((featureFlags & License.REMOTE_SAM_CLNT_SUPPORT)
            == License.REMOTE_SAM_CLNT_SUPPORT) ? true : false);
    }

    public boolean isMigrationToolkitSupported() {
        return (((featureFlags & License.MIG_TOOLKIT_SUPPORT)
            == License.MIG_TOOLKIT_SUPPORT) ? true : false);
    }

    public boolean isFastFileSystemSupported() {
        return (((featureFlags & License.FAST_FS_SUPPORT)
            == License.FAST_FS_SUPPORT) ? true : false);
    }

    public boolean isDatabaseSupported() {
        return (((featureFlags & License.DATABASE_SUPPORT)
            == License.DATABASE_SUPPORT) ? true : false);
    }

    public boolean isForeignTapeSupported() {
        return (((featureFlags & License.FOREIGN_TAPE_SUPPORT)
            == License.FOREIGN_TAPE_SUPPORT) ? true : false);
    }

    public boolean isSegmentSupported() {
        return (((featureFlags & License.SEGMENT_SUPPORT)
            == License.SEGMENT_SUPPORT) ? true : false);
    }

    public boolean isSharedFileSystemSupported() {
        return (((featureFlags & License.SHARED_FS_SUPPORT)
            == License.SHARED_FS_SUPPORT) ? true : false);
    }

    public boolean isSANAPISupported() {
        return (((featureFlags & License.SANAPI_SUPPORT)
            == License.SANAPI_SUPPORT) ? true : false);
    }

    public boolean isStandaloneQFSSupported() {
        return (((featureFlags & License.STANDALONE_QFS_SUPPORT)
            == License.STANDALONE_QFS_SUPPORT) ? true : false);
    }

    public MdLicense[] getMediaLicenses() { return mdLicenses; }

    public String toString() {
        String s = "License Type: " + fsType + " " + type + "," +
        " Expiration: " + expire +
        " Host ID: " + Long.toHexString(hostID) +
        "\nRemote server: " + isRemoteSamServerSupported() +
        "\nRemote client: " + isRemoteSamClientSupported() +
        "\nMigration: " + isMigrationToolkitSupported() +
        "\nQFS: " + isStandaloneQFSSupported() +
        "\nData Base features: " + isDatabaseSupported() +
        "\nForeign tape support: " + isForeignTapeSupported() +
        "\nSAN API support: " + isSANAPISupported() +
        "\nSegment feature: " + isSegmentSupported() +
        "\nShared filesystem support: " + isSharedFileSystemSupported() +
        "\nFast filesystem support: " + isFastFileSystemSupported() + "\n";

        if (mdLicenses != null) {
          for (int i = 0; i < mdLicenses.length; i++) {
              s += "   " + mdLicenses[i] + "\n";
          }
        } else {
              s += "  none\n";
        }

        return s;
    }

    public static native short getFSType(Ctx x) throws SamFSException;

    public static native License getLicense(Ctx c) throws SamFSException;
}

File: src/lib/sammgmt/license.c
/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */
#pragma ident   "$Revision: 1.24 $"

static char *_SrcFile = __FILE__;  /* Using __FILE__ makes duplicate strings */

#include <limits.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

#include "sam/lib.h"
#include "sam/types.h"
#include "sam/sam_trace.h"
#include "sam/syscall.h"
#include "aml/proto.h"

#include "mgmt/util.h"
#include "mgmt/config/common.h"
#include "pub/mgmt/error.h"
#include "pub/mgmt/license.h"

/*
 * Get the license type:
 *	 NON_EXPIRING, EXPIRING, DEMO, QFS_TRIAL, or QFS_SPECIAL
 */
int
get_license_type(
ctx_t *ctx /* ARGSUSED */)
{
	return (NON_EXPIRING);
}


/*
 * Get the license expiration date if the license type is EXPIRING
 * (in seconds since 1970)
 */
int
get_expiration_date(
ctx_t *ctx /* ARGSUSED */)
{
	return (0);
}


/*
 * Get the SAMFS type: QFS, SAM-FS, or SAM-QFS
 * used by the filesystem api. Only do the stat one time.
 */
int
get_samfs_type(
ctx_t *ctx /* ARGSUSED */)
{
	static int type = 0;
	struct stat st;

	if (type != 0) {
		return (type);
	} else if (stat(SBIN_DIR"/"SAM_AMLD, &st) == 0) {
		type = SAMQFS;
	} else {
		type = QFS;
	}

	return (type);
}

/*
 * Get the total number of licensed slots for the given
 * robot_type and media_type pair
 */
int
get_licensed_slots(
ctx_t *ctx,			/* ARGSUSED */
ushort_t robot_type,	/* IN - robot type */
ushort_t media_type)	/* IN - media type */
{
	/* INT32_MAX = 0x7fffffff, MAX_SLOTS = 0x7fffffff */
	return (MAX_SLOTS - 1);

}


/*
 * Get the total number of licensed slots for the given
 * robot_type and media_type pair
 */
int
get_licensed_media_slots(
ctx_t *ctx,				/* ARGSUSED */
const char *robot_name,	/* IN - robot name */
const char *media_name)	/* IN - media name */
{
	return (MAX_SLOTS - 1);
}


/*
 * Get a list of licensed media_types (strings)
 */
int
get_licensed_media_types(
ctx_t *ctx,				/* ARGSUSED */
sqm_lst_t **media_list)	/* OUT - list of licensed media names */
{

	sqm_lst_t *lst = NULL;

	lst = lst_create();

	if (!lst) {
		Trace(TR_OPRMSG, "out of memory");
		return (-1);
	}

	*media_list = lst;

	return (lst->length);

}

/*
 * Get the license information:
 *	 license type, hostid, features supported, media licenses
 */
int
get_license_info(
ctx_t *ctx, 			/* ARGSUSED */
license_info_t **info		/* OUT - License info struct */
)
{

	struct stat st;

	Trace(TR_MISC, "getting license information");

	if (ISNULL(info)) {
		Trace(TR_OPRMSG, "null argument found");
		/* samerrno and samerrmsg are set */
		return (-1);
	}

	*info = NULL;

	/* fill in the license_info structure */

	*info = (license_info_t *)mallocer(sizeof (license_info_t));

	if (*info == NULL) {
		Trace(TR_ERR, "get license info failed: %s", samerrmsg);
		return (-1);
	}

	memset(*info, 0, sizeof (license_info_t));

	(*info)->type = NON_EXPIRING;
	(*info)->expire = 0;
	(*info)->hostid = 0;

	if (stat(SBIN_DIR"/"SAM_AMLD, &st) == 0) {
		(*info)->fsType = SAMQFS;
	} else {
		(*info)->fsType = QFS;
	}

	/* clear the bitmask */
	(*info)->feature_flags = (uint32_t)0;

	(*info)->feature_flags |= 0xffffffff;

	return (0);

}
